### ISSUE TITLE
Add Test.todo (was: add `pending` expectation)

### ISSUE THREAD

**Issue by soenkehahn** [elm-community/elm-test#94]
`pending : Expectation` would have similar behavior as `pass : Expectation`, but test-runners would report the test-case not as passing or failing, but as pending.

This is useful during development to jot down test-cases (with descriptions) without having to implement them immediately, e.g:

``` elm
all : Test
all =
  describe "myNewFunction"
    [ test "behaves correct in the common case" myAwesomeTest
    , test "a corner-case that I thought about" pending
    , test "another corner-case" pending
    ]
```

This would be similar to https://www.stackage.org/haddock/lts-7.14/hspec-2.2.4/Test-Hspec.html#v:pending.

**Comment by rtfeldman** [comment]
Hm...why would this be better than `Expect.fail "pending"`? The latter would encourage finishing it. 😄

**Comment by soenkehahn** [comment]
The difference in workflow is not big, but I think `pending` is nice to have.

Consider the following scenario: I'm adding a new, fairly complex feature using TDD. I've added a bunch of pending tests with `fail "pending"`. Now I pick one of those tests and implement it. Now the output of my test-suite is cluttered with all the remaining `pending` fails, although at the moment I'm not interested in pending test-cases, but would much rather clearly see the output of the one failing test-case that I already implemented. To be able to focus on fixing that one failing test.

Furthermore, if I start changing the production code to make that one test-case pass, I might have a regression that's discovered by some other test I wrote before. Again I would want this test-failure to be very obvious in the test-suite output and not to be interspersed with the noise of failing `pending` test-cases.

Another workaround that doesn't suffer from this is to use `pass` and add source code comments e.g. `-- fixme: pending`. Then I can `grep` for `pending` (or `fixme`) in the code to see whether there are pending tests. This is what I'm doing right now -- but having `pending` would be nicer.

I tried to find some nice write-up about how to use pending and why it's useful. The best I could find is this: http://stackoverflow.com/questions/8017514/tdd-practice-distinguishing-between-genuine-failures-and-unimplemented-features

**Comment by rtfeldman** [comment]
Ah, that's interesting. What if pending tests fail the test run, but don't display per-failure output? 🤔

**Comment by mgold** [comment]
To Richard's first comment, the actual syntactical noise comparison is:
```elm
, test "description" <| \_ -> Expect.fail "pending"
, test "description" Expect.pending
-- assuming
Expect.pending : a -> Expectation
```

Defining the function this way that it is for a stubbed out test. (I'd avoid `Expect.stubbed` because stubbing has a different meaning in testing contexts. "Pending" is the accepted word.)

One of the few times I can remember committing a pending tests (in Rails/Rspec, at a job) was testing a feature that was enabled with a flag. If the flag was active, testing could proceed; if not no testing could be done so I marked it as pending. Because Elm doesn't let you look at global read-only state with the ease of Rails, I think we can say this case won't happen (very often) and that marking a test as pending *conditionally* is a bad idea.

Rspec treats pending tests the opposite of this proposal: they don't cause the suite to fail but do cause an error message. I agree with the justification for the proposed behavior, and it might make more sense in the context of a statically typed language. We will, of course, need to document it.

The other professional experience with pending tests that I remember is a colleague seeing pending tests, furrowing his brow, and muttering "I like not that". So that's an anecdote in favor of pending means failed, so these tests of unknown value or intent don't get forgotten about.

I like the direction this is going!

**Comment by rtfeldman** [comment]
For that use case, I think this type signature makes more sense:

```elm
Test.pending : (a -> Expectation) -> a -> Expectation
```

This way you can actually write out the real test, including the real expectation, and mark it as pending until you're ready to verify its implementation.

```elm
    import Json.Decode exposing (decodeString, int)
    import Test exposing (test)
    import Expect


    test "Json.Decode.int can decode the number 42." <|
        Test.pending <|
            \() ->
                case decodeString int "42" of
                    Ok _ ->
                        Expect.pass

                    Err err ->
                        Expect.fail err
```

Thoughts?

**Comment by rtfeldman** [comment]
Thinking about this some more, I'm doubting this is a good idea after all.

If I'm doing test-first, it's important that I start off with genuine failures for all my tests - *actual* red tests, not just "red because pending" - otherwise, for all I know they might pass in all cases and be incapable of failing. If I use `pending` at this stage, I'm potentially masking problems in my tests.

So assuming `pending` exists, and I want to use it, my workflow ought to go like this:

1. Write a bunch of actually failing tests
2. Run all my tests to make sure it's 100% failures, no passes
3. Go through and mark everything as `pending` to hide the failure messages
4. Remove `pending` from things when I think they're done, to confirm that they turn green

It seems like in the typical case, I'm creating more work for myself—going through and adding `pending` and then removing it piecemeal afterward. The alternative seems easier: load up on failures and watch them disappear as I implement things, with the failures that remain telling me what I have left to do.

If I find a particular failure difficult to resolve, wouldn't the right course of action be to filter my tests such that only that one runs? (Granted, `Test.filter` hasn't been translated into a `--filter` CLI flag yet, but I intend to add that features to the CLI to make it easier to run just one test.)

Regarding [the aforementioned StackOverflow question](http://stackoverflow.com/questions/8017514/tdd-practice-distinguishing-between-genuine-failures-and-unimplemented-features):

> If you are in the middle of a TDD iteration, how do you know which tests fail because the existing code is genuinely incorrect and which fail because either the test itself or the features haven't been implemented yet? Please don't say, "you just don't care, because you have to fix both." I'm ready to move past that mindset.

The OP anticipates a sensible response and rejects it without giving a reason. I can't justify feature requests with tautologies - "I don't want to do that because I don't want to do that."

As for wanting to mark some things as pending because I'm only partway through implementing all my tests...under those circumstances, what's the point of running *any* tests? Periodically checking that what I've written compiles seems like it would serve the purpose of breaking the work into manageable chunks. Running the tests and confirming they're all failures (or "pending") doesn't seem like it adds value over compilation. (What significant information does that tell me that I couldn't have gotten using existing tools?)

If my goal is to focus on getting a few tests to pass, even though the others haven't even been fully written yet...once again, filtering seems like a better tool for this.

Basically upon further thought I'm not convinced there's a good niche for this after all.

**Comment by mgold** [comment]
I don't think the intention is to write out many test cases that are all marked as pending. My understanding of TDD is to write *one failing test at a time*. Because it's just as easy for the test to be wrong as the code to be wrong, you avoid duplicating the same bug or flaw over many tests by making sure you only copy a working test.

*But*, what happens when you have many ideas for tests but can't focus on them all at once? You want to write them down and not forget to come back to them. This is where pending tests come in. In Rspec, you can even do it implicitly:

```ruby
it "reticulates splines" do
  expect(spline.reticulate).to be_reticulated
end

it "does not deticulate a pfrine"
```

The first test is a normal implemented test, with `do...end` being the equivalent of elm-test's lambdas. The second test is not implemented, and this will default to pending (reason: not implemented). It's reasonable for the developer to have started by writing many tests as not implemented stubs, and now (s)he's partway through implementing the suite.

If we want to encourage only this use-case, maybe a careful choice of API and name will help:

```elm
, test "addition commutes" <| \() -> 1 + 2 |> Expect.equal (2+1)
, todo "addition associates"
, todo "multiplication commutes"

-- because

Test.todo : String -> Test

-- the alternative is

, test "addition commutes" <| \() -> 1 + 2 |> Expect.equal (2+1)
-- TODO: addition associates
-- TODO: multiplication commutes
```

I think this API is less likely to be abused (you can't use it to create a test that is sometimes pending), and it should be fairly easy to implement as a failing test whose `Reason` is a new tag. TODO tests would cause a failure but (by default) the runners would not print their error messages, beyond "x tests TODO not shown". The logic is the same as above; we know these tests failed so please don't clutter the output as we look for failures of the test we're currently implementing, or regressions found by other tests.

**Comment by rtfeldman** [comment]
Aha! I dig that.

Also we could hide the TODO tests from the output when you have non-TODO failures (so they don't clutter up the output), but once you have fixed all non-TODO failures, and only TODOs remain, we could print out the TODOs like so:

```
TEST RUN FAILED

There are 2 remaining TODO tests:

• addition commutes
• multiplication commutes
```

**Comment by mgold** [comment]
Oh I like that too! I'll take a crack at implementation when I get a chance.

**Comment by soenkehahn** [comment]
Awesome to see someone working on this, thanks @mgold! I also really like the type `String -> Test`. Makes it even faster to scribble down a `todo` test.

**Comment by mgold** [comment]
Yeah, the best ideas rarely emerge fully-formed. We don't want to copy Rspec/Jasmine/etc just because they do something that way.
